chunky/src/java/se/llbit/math/Grid.java

```diff
@@ -47,32 +46,56 @@ public EmitterPosition(int x, int y, int z) {
   // This way of storing the data is more difficult to manipulate but more
   // memory efficient by virtue of only having 2 flat arrays
 
-  public Grid(int octreeDepth, int cellSize) {
-    this.cellSize = cellSize;
-    long totalSize = (1L << octreeDepth);
-    this.gridSize = (int) ((totalSize + (cellSize-1)) / cellSize);
-  }
+  // Only used during construction
+  private int minX, maxX, minY, maxY, minZ, maxZ;
+
+  private int offsetX, sizeX, offsetY, sizeY, offsetZ, sizeZ;
 
-  // Constructor used by the load method
-  private Grid(int gridSize, int cellSize, int overloadSelectFlag /*unused*/) {
-    this.gridSize = gridSize;
+  public Grid(int cellSize) {
     this.cellSize = cellSize;
+    minX = maxX = minY = maxY = minZ = maxZ = -1;
   }
 
   public void addEmitter(int x, int y, int z) {
     emitterPositions.add(new EmitterPosition(x, y, z));
+    if(minX == -1 || x < minX)
+      minX = x;
+    if(maxX == -1 || x  > maxX)
+      maxX = x;
+    if(minY == -1 || y < minY)
+      minY = y;
+    if(maxY == -1 || y > maxY)
+      maxY = y;
+    if(minZ == -1 || z < minZ)
+      minZ = z;
+    if(maxZ == -1 || z > maxZ)
+      maxZ = z;
   }
 
   private int cellIndex(int x, int y, int z) {
-    return ((y * gridSize) + x) * gridSize + z;
+    return (((y - offsetY) * sizeX) + (x - offsetX)) * sizeZ + (z - offsetZ);
+  }
+
+  private boolean isOutOfBounds(int x, int y, int z)
+  {
+    return x < offsetX || x >= sizeX + offsetX
+            || y < offsetY || y >= sizeY + offsetY
+            || z < offsetZ || z >= sizeZ + offsetZ;
   }
 
   /**
    * Needs to be call when every emitter has been added
```

Comment:
:thinking: Maybe we should add a `GridBuilder` to enforce this? That would also allow to drop minXYZ and maxXYZ from the grid itself. But that's just minor refactoring, nothing to worry about in this PR.